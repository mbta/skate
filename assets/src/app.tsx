// We need to import the CSS so that webpack will load it.
// The MiniCssExtractPlugin is used to separate it out into
// its own CSS file.
declare function require(name: string): string
// tslint:disable-next-line
require("../css/app.scss")

// webpack automatically bundles all modules in your
// entry points. Those entry points can be configured
// in "webpack.config.js".
//
// Import dependencies
//
import "@babel/polyfill"
import "leaflet-defaulticon-compatibility/dist/leaflet-defaulticon-compatibility.webpack.css" // see https://github.com/Leaflet/Leaflet/issues/4968#issuecomment-483402699
import "leaflet/dist/leaflet.css"
import "phoenix_html"
import * as React from "react"
import ReactDOM from "react-dom"
import AppStateWrapper from "./components/appStateWrapper"

if (window.FS && window.username) {
  window.FS.identify(window.username, { displayName: window.username })
}

// Import local files
//
// Local files can be imported directly using relative paths, for example:
// import socket from "./socket"

ReactDOM.render(<AppStateWrapper />, document.getElementById("app"))
