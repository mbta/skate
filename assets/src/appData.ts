const appData = (): DOMStringMap | undefined => {
  const dataEl = document.getElementById("app")

  if (!dataEl) {
    return undefined
  }

  return dataEl.dataset
}

export default appData
