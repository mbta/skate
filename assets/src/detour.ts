import { LatLngLiteral } from "leaflet"
import { Shape, ShapePoint } from "./schedule"

export interface DetourShape {
  coordinates: ShapePoint[]
  directions: {
    instruction: string
  }[]
}

export interface OriginalRoute {
  routeName: string
  routeDescription: string
  routeOrigin: string
  routeDirection: string
  shape: Shape
  center: LatLngLiteral
  zoom: number
}
