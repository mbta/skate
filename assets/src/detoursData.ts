import { array, Infer, number, string, type } from "superstruct"
import { DraftDetours } from "./models/detours"

export const DraftDetoursData = array(
  type({
    route: string(),
    direction: string(),
    name: string(),
    intersection: string(),
    active_since: number(),
  })
)
export type DraftDetoursData = Infer<typeof DraftDetoursData>

export const draftDetoursFromData = (
  draftDetoursData: DraftDetoursData
): DraftDetours =>
  draftDetoursData.map((draft) => ({
    route: draft.route,
    direction: draft.direction,
    name: draft.name,
    intersection: draft.intersection,
    activeSince: draft.active_since,
  }))
