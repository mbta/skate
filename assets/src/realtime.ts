import {
  BlockId,
  DirectionId,
  RouteId,
  RoutePatternId,
  StopId,
  TimepointId,
  TripId,
  ViaVariant,
} from "./schedule"

import { Crowding } from "./models/crowding"
import { DetourId } from "./models/detoursList"

export interface BlockWaiver {
  startTime: Date
  endTime: Date
  causeId: number
  causeDescription: string
  remark: string | null
}

export interface DataDiscrepancy {
  attribute: string
  sources: DataDiscrepancySource[]
}

export interface DataDiscrepancySource {
  id: SourceId
  value: string | null
}

export type EndOfTripType = "another_trip" | "swing_off" | "pull_back"

export interface Ghost {
  id: VehicleId
  directionId: DirectionId
  routeId: RouteId
  routePatternId: RoutePatternId | null
  tripId: TripId
  headsign: string
  blockId: BlockId
  runId: RunId | null
  viaVariant: ViaVariant | null
  incomingTripDirectionId: DirectionId | null
  layoverDepartureTime: number | null
  scheduledTimepointStatus: VehicleTimepointStatus | null
  scheduledLogonTime: number | null
  routeStatus: RouteStatus
  blockWaivers: BlockWaiver[]
  currentPieceStartPlace: string | null
  currentPieceFirstRoute: RouteId | null
}

export type NotificationId = string

// A Notification's `content` is populated with an Elixir struct and the
// `__struct__` field is serialized so that the type can be discerned on the
// frontend.
export enum NotificationType {
  BridgeMovement = "Elixir.Notifications.Db.BridgeMovement",
  BlockWaiver = "Elixir.Notifications.Db.BlockWaiver",
  Detour = "Elixir.Notifications.Db.Detour",
}

export interface BridgeLoweredNotification {
  $type: NotificationType.BridgeMovement
  status: "lowered"
}

export interface BridgeRaisedNotification {
  $type: NotificationType.BridgeMovement
  status: "raised"
  loweringTime: Date
}

export type BridgeNotification =
  | BridgeLoweredNotification
  | BridgeRaisedNotification

export type BlockWaiverNotification = {
  $type: NotificationType.BlockWaiver
  createdAt: Date
  reason: BlockWaiverReason
  routeIds: RouteId[]
  runIds: RunId[]
  tripIds: TripId[]
  operatorId: string | null
  operatorName: string | null
  routeIdAtCreation: string | null
  startTime: Date
  endTime: Date | null
}

export enum DetourNotificationStatus {
  Activated = "activated",
  Deactivated = "deactivated",
}

export type DetourNotification = {
  $type: NotificationType.Detour
  status: DetourNotificationStatus
  detourId: DetourId
  headsign: string
  route: string
  direction: string
  origin: string
}

export type NotificationContentTypes =
  | BridgeNotification
  | BlockWaiverNotification
  | DetourNotification

export interface Notification<
  TNotification extends NotificationContentTypes = NotificationContentTypes
> {
  id: NotificationId
  createdAt: Date
  state: NotificationState
  content: TNotification
}

export function isBlockWaiverNotification(
  notification: Notification
): notification is Notification<BlockWaiverNotification> {
  return notification.content.$type === NotificationType.BlockWaiver
}
export type BlockWaiverReason =
  | "other"
  | "accident"
  | "adjusted"
  | "disabled"
  | "diverted"
  | "manpower"
  | "operator_error"
  | "traffic"

export type NotificationState = "unread" | "read" | "deleted"

export type SourceId = string

export type RouteStatus = "on_route" | "laying_over" | "pulling_out"

export type RunId = string

export interface ByRunId<T> {
  [runId: string]: T
}

export type TrainVehicleId = string

export interface TrainVehicle {
  id: TrainVehicleId
  latitude: number
  longitude: number
  bearing: number
}

export interface Vehicle {
  id: VehicleId
  label: string | null
  runId: RunId | null
  timestamp: number
  latitude: number
  longitude: number
  directionId: DirectionId | null
  routeId: RouteId | null
  routePatternId: RoutePatternId | null
  tripId: TripId | null
  headsign: string | null
  viaVariant: ViaVariant | null
  operatorId: string | null
  operatorFirstName: string | null
  operatorLastName: string | null
  operatorLogonTime: Date | null
  overloadOffset?: number
  bearing?: number
  blockId: BlockId | null
  previousVehicleId: string | null
  scheduleAdherenceSecs: number | null
  incomingTripDirectionId: DirectionId | null
  isShuttle: boolean
  isOverload: boolean
  isOffCourse: boolean
  isRevenue: boolean | null
  layoverDepartureTime: number | null
  pullbackPlaceName: string | null
  dataDiscrepancies: DataDiscrepancy[]
  stopStatus: VehicleStopStatus
  timepointStatus: VehicleTimepointStatus | null
  scheduledLocation: VehicleScheduledLocation | null
  routeStatus: RouteStatus
  endOfTripType: EndOfTripType
  blockWaivers: BlockWaiver[]
  crowding: Crowding | null
}

export interface VehicleInScheduledService extends Vehicle {
  directionId: DirectionId
  routeId: RouteId
  blockId: BlockId
}

export type VehicleId = string

export interface VehicleStopStatus {
  stopId: StopId | null
  stopName: string | null
}

export interface VehicleScheduledLocation {
  routeId: RouteId
  routePatternId: RoutePatternId | null
  directionId: DirectionId
  tripId: TripId
  runId: RunId
  timeSinceTripStartTime: number
  headsign: string | null
  viaVariant: ViaVariant | null
  timepointStatus: VehicleTimepointStatus
}

export interface VehicleTimepointStatus {
  timepointId: TimepointId
  fractionUntilTimepoint: number
}
