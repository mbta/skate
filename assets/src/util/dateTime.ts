export const now = (): Date => new Date()

export const dateFromEpochSeconds = (time: number): Date =>
  new Date(time * 1_000)

export const formattedTime = (date: Date): string => {
  const hours24 = date.getHours()
  return `${hours12(hours24)}:${zeroPad(date.getMinutes())}${ampm(hours24)}`
}

export const formattedTimeDiff = (a: Date, b: Date): string => {
  const diffMs: number = a.valueOf() - b.valueOf()
  const diffHours = Math.floor(diffMs / 3_600_000)
  const diffMinutes = Math.floor((diffMs % 3_600_000) / 60_000)

  const diffMinutesStr = `${diffMinutes}m`

  return diffHours >= 1 ? `${diffHours}h ${diffMinutesStr}` : diffMinutesStr
}

/** Takes a time of day in seconds since midnight
 */
export const formattedScheduledTime = (time: number): string => {
  const minutes = Math.floor(time / 60)
  const hours25 = Math.floor(minutes / 60)
  const minutes60 = minutes - hours25 * 60
  return `${hours12(hours25)}:${zeroPad(minutes60)}${ampm(hours25)}`
}

export const hours12 = (hours25: number): number => hours25 % 12 || 12

export const ampm = (hours24: number): string =>
  hours24 >= 12 && hours24 < 24 ? "pm" : "am"

const zeroPad = (time: number): string => (time < 10 ? `0${time}` : `${time}`)
