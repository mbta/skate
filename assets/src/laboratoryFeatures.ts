import appData from "./appData"

const featureIsEnabled = (key: string): boolean => {
  const data = appData()

  if (data === undefined || data.laboratoryFeatures === undefined) {
    return false
  }

  const features = JSON.parse(data.laboratoryFeatures)
  if (features[key] === undefined) {
    return false
  }

  return features[key]
}

export default featureIsEnabled
