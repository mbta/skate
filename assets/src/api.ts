import "whatwg-fetch"
import appData from "./appData"
import { Block, Run } from "./minischedule"
import { reload } from "./models/browser"
import { blockFromData, runFromData } from "./models/minischeduleData"
import { swingsFromData } from "./models/swingsData"
import { NotificationId, NotificationState } from "./realtime.d"
import { RouteSettings } from "./routeSettings"
import {
  DirectionName,
  Route,
  RouteId,
  Shape,
  Swing,
  Timepoint,
  TripId,
} from "./schedule.d"

export interface RouteData {
  id: string
  direction_names: {
    "0": DirectionName
    "1": DirectionName
  }
  name: string
}

const checkResponseStatus = (response: Response) => {
  if (response.status === 200) {
    return response
  }

  if (Math.floor(response.status / 100) === 3 || response.status === 403) {
    // If the API sends us a redirect or forbidden, the user needs to
    // re-authenticate. Reload to go through the auth flow again.
    reload(true)
  }

  throw new Error(`Response error: ${response.status}`)
}

const parseJson = (response: Response) => response.json()

export const apiCall = <T>({
  url,
  parser,
  defaultResult,
}: {
  url: string
  parser: (data: any) => T
  defaultResult?: T
}): Promise<T> =>
  fetch(url)
    .then(checkResponseStatus)
    .then(parseJson)
    .then(({ data: data }: { data: any }) => parser(data))
    .catch((error) => {
      if (defaultResult === undefined) {
        throw error
      } else {
        return defaultResult
      }
    })

export const parseRouteData = ({
  id,
  direction_names,
  name,
}: RouteData): Route => ({
  id,
  directionNames: direction_names,
  name,
})

const parseRoutesData = (routesData: RouteData[]): Route[] =>
  routesData.map(parseRouteData)

export const fetchRoutes = (): Promise<Route[]> =>
  apiCall({
    url: "/api/routes",
    parser: parseRoutesData,
  })

export const fetchShapeForRoute = (routeId: RouteId): Promise<Shape[]> =>
  apiCall({
    url: `/api/shapes/route/${routeId}`,
    parser: (shapes: Shape[]) => shapes,
    defaultResult: [],
  })

export const fetchShapeForTrip = (tripId: TripId): Promise<Shape | null> =>
  apiCall({
    url: `/api/shapes/trip/${tripId}`,
    parser: (shape: Shape | null) => shape,
    defaultResult: null,
  })

export const fetchShuttleRoutes = (): Promise<Route[]> =>
  apiCall({
    url: "/api/shuttles",
    parser: parseRoutesData,
    defaultResult: [],
  })

export const fetchTimepointsForRoute = (
  routeId: RouteId
): Promise<Timepoint[]> =>
  apiCall({
    url: `/api/routes/${routeId}`,
    parser: (timepoints: Timepoint[]) => timepoints,
    defaultResult: [],
  })

export const fetchMinischeduleRun = (tripId: TripId): Promise<Run | null> =>
  apiCall({
    url: `/api/minischedule/run/${tripId}`,
    parser: nullableParser(runFromData),
    defaultResult: null,
  })

export const fetchMinischeduleBlock = (tripId: TripId): Promise<Block | null> =>
  apiCall({
    url: `/api/minischedule/block/${tripId}`,
    parser: nullableParser(blockFromData),
    defaultResult: null,
  })

export const fetchNearestIntersection = (
  latitude: number,
  longitude: number
): Promise<string | null> =>
  apiCall({
    url: `/api/intersection?latitude=${latitude}&longitude=${longitude}`,
    parser: nullableParser((intersection: string) => intersection),
    defaultResult: null,
  })

export const fetchSwings = (): Promise<Swing[] | null> =>
  apiCall({
    url: "/api/swings",
    parser: nullableParser(swingsFromData),
    defaultResult: null,
  })

export const putNotificationReadState = (
  newReadState: NotificationState,
  notificationIds: NotificationId[]
): void => {
  const url = `/api/notification_read_state?new_state=${newReadState}&notification_ids=${notificationIds.join(
    ","
  )}`
  fetch(url, {
    method: "PUT",
    headers: {
      "Content-Type": "application/json",
      "x-csrf-token": getCsrfToken(),
    },
  })
}

export const putUserSetting = (field: string, value: string): void => {
  const url = `/api/user_settings?field=${field}&value=${value}`
  fetch(url, {
    method: "PUT",
    headers: {
      "Content-Type": "application/json",
      "x-csrf-token": getCsrfToken(),
    },
  })
}

export const putRouteSettings = (routeSettings: RouteSettings): void => {
  fetch("/api/route_settings", {
    method: "PUT",
    headers: {
      "Content-Type": "application/json",
      "x-csrf-token": getCsrfToken(),
    },
    body: JSON.stringify(routeSettings),
  })
}

const getCsrfToken = (): string => appData()?.csrfToken || ""

export const nullableParser =
  <Data, T>(parser: (data: Data) => T): ((data: Data | null) => T | null) =>
  (data: Data | null) =>
    data === null ? null : parser(data)
