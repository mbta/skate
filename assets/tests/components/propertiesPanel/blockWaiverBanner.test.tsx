import React from "react"
import renderer from "react-test-renderer"
import BlockWaiverBanner from "../../../src/components/propertiesPanel/blockWaiverBanner"
import { BlockWaiver } from "../../../src/realtime"
import * as dateTime from "../../../src/util/dateTime"

describe("BlockWaiverBanner", () => {
  jest
    .spyOn(dateTime, "now")
    .mockImplementation(() => new Date("2020-02-25T16:10:00.000Z"))

  test("renders a current time waiver", () => {
    const blockWaiver: BlockWaiver = {
      startTime: new Date("2020-02-25T15:53:20.000Z"),
      endTime: new Date("2020-02-25T16:26:40.000Z"),
      remark: "E:1106",
    }
    const tree = renderer
      .create(<BlockWaiverBanner blockWaiver={blockWaiver} />)
      .toJSON()

    expect(tree).toMatchSnapshot()
  })

  test("renders a future time waiver", () => {
    const blockWaiver: BlockWaiver = {
      startTime: new Date("2020-02-25T16:26:40.000Z"),
      endTime: new Date("2020-02-25T16:43:20.000Z"),
      remark: "E:1106",
    }
    const tree = renderer
      .create(<BlockWaiverBanner blockWaiver={blockWaiver} />)
      .toJSON()

    expect(tree).toMatchSnapshot()
  })

  test("renders a past time waiver", () => {
    const blockWaiver: BlockWaiver = {
      startTime: new Date("2020-02-25T15:36:40.000Z"),
      endTime: new Date("2020-02-25T15:53:20.000Z"),
      remark: "E:1106",
    }
    const tree = renderer
      .create(<BlockWaiverBanner blockWaiver={blockWaiver} />)
      .toJSON()

    expect(tree).toMatchSnapshot()
  })
})
