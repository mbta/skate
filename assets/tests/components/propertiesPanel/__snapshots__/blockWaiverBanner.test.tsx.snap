// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BlockWaiverBanner renders a current time waiver 1`] = `
<div
  className="m-block-waiver-banner m-block-waiver-banner--current"
>
  <div
    className="m-block-waiver-banner__header"
  >
    <span
      className="m-block-waiver-banner__alert-icon"
    >
      <svg
        className="c-icon-alert-circle"
        viewBox="-3 -3 54 54"
      >
        <g
          className="c-icon-alert-circle--black"
        >
          <circle
            className="c-icon-alert-circle__outline"
            cx="24"
            cy="24"
            r="27"
          />
          <circle
            className="c-icon-alert-circle__fill"
            cx="24"
            cy="24"
            r="22.59"
          />
          <g
            className="c-icon-alert-circle__exclamation-point"
          >
            <path
              d="m20.39 4.42h7.22a1.81 1.81 0 0 1 1.49 2.11l-1.49 23.22a1.73 1.73 0 0 1 -1.49 1.78h-4.24a1.73 1.73 0 0 1 -1.49-1.78l-1.49-23.22c-.07-1.13.61-2.11 1.49-2.11"
            />
            <path
              d="m21.57 34.54h4.87a1.88 1.88 0 0 1 1.82 1.93v5.17a1.89 1.89 0 0 1 -1.82 1.94h-4.87a1.89 1.89 0 0 1 -1.83-1.94v-5.17a1.88 1.88 0 0 1 1.83-1.93"
            />
          </g>
        </g>
      </svg>
    </span>
    <div
      className="m-block-waiver-banner__title"
    >
      Dispatcher Note - 
      Current
    </div>
  </div>
  <table
    className="m-block-waiver-banner__details"
  >
    <tbody>
      <tr>
        <td
          className="m-block-waiver-banner__detail-label m-block-waiver-banner__reason"
          data-event="click"
          data-for="reason"
          data-tip={true}
        >
          Reason 
          <span
            className="m-block-waiver-banner__reason-icon"
            dangerouslySetInnerHTML={
              Object {
                "__html": "SVG",
              }
            }
          />
        </td>
        <td
          className="m-block-waiver-banner__detail-value"
        >
          E:1106
        </td>
      </tr>
      <tr>
        <td
          className="m-block-waiver-banner__detail-label m-block-waiver-banner__detail-label--start-time"
        >
          Start Time
        </td>
        <td
          className="m-block-waiver-banner__detail-value m-block-waiver-banner__detail-value--start-time"
        >
          3:53pm
        </td>
      </tr>
      <tr>
        <td
          className="m-block-waiver-banner__detail-label m-block-waiver-banner__detail-label--end-time"
        >
          End Time
        </td>
        <td
          className="m-block-waiver-banner__detail-value m-block-waiver-banner__detail-value--end-time"
        >
          4:26pm
        </td>
      </tr>
    </tbody>
  </table>
  <div
    className="__react_component_tooltip-304517788 __react_component_tooltip place-bottom type-dark"
    data-id="tooltip"
    id="reason"
  >
    <h3>
      Reason codes
    </h3>
    <ul>
      <li>
        <strong>
          B
        </strong>
         - Manpower
      </li>
      <li>
        <strong>
          C
        </strong>
         - No equipment
      </li>
      <li>
        <strong>
          D
        </strong>
         - Disabled bus
      </li>
      <li>
        <strong>
          E
        </strong>
         - Diverted to other work
      </li>
      <li>
        <strong>
          F
        </strong>
         - Traffic
      </li>
      <li>
        <strong>
          G
        </strong>
         - Accident
      </li>
      <li>
        <strong>
          H
        </strong>
         - Weather
      </li>
      <li>
        <strong>
          I
        </strong>
         - Operator error
      </li>
      <li>
        <strong>
          J
        </strong>
         - Other
      </li>
    </ul>
  </div>
</div>
`;

exports[`BlockWaiverBanner renders a future time waiver 1`] = `
<div
  className="m-block-waiver-banner m-block-waiver-banner--future"
>
  <div
    className="m-block-waiver-banner__header"
  >
    <span
      className="m-block-waiver-banner__alert-icon"
    >
      <svg
        className="c-icon-alert-circle"
        viewBox="-3 -3 54 54"
      >
        <g
          className="c-icon-alert-circle--grey-on-grey"
        >
          <circle
            className="c-icon-alert-circle__outline"
            cx="24"
            cy="24"
            r="27"
          />
          <circle
            className="c-icon-alert-circle__fill"
            cx="24"
            cy="24"
            r="22.59"
          />
          <g
            className="c-icon-alert-circle__exclamation-point"
          >
            <path
              d="m20.39 4.42h7.22a1.81 1.81 0 0 1 1.49 2.11l-1.49 23.22a1.73 1.73 0 0 1 -1.49 1.78h-4.24a1.73 1.73 0 0 1 -1.49-1.78l-1.49-23.22c-.07-1.13.61-2.11 1.49-2.11"
            />
            <path
              d="m21.57 34.54h4.87a1.88 1.88 0 0 1 1.82 1.93v5.17a1.89 1.89 0 0 1 -1.82 1.94h-4.87a1.89 1.89 0 0 1 -1.83-1.94v-5.17a1.88 1.88 0 0 1 1.83-1.93"
            />
          </g>
        </g>
      </svg>
    </span>
    <div
      className="m-block-waiver-banner__title"
    >
      Dispatcher Note - 
      Future
    </div>
  </div>
  <table
    className="m-block-waiver-banner__details"
  >
    <tbody>
      <tr>
        <td
          className="m-block-waiver-banner__detail-label m-block-waiver-banner__reason"
          data-event="click"
          data-for="reason"
          data-tip={true}
        >
          Reason 
          <span
            className="m-block-waiver-banner__reason-icon"
            dangerouslySetInnerHTML={
              Object {
                "__html": "SVG",
              }
            }
          />
        </td>
        <td
          className="m-block-waiver-banner__detail-value"
        >
          E:1106
        </td>
      </tr>
      <tr>
        <td
          className="m-block-waiver-banner__detail-label m-block-waiver-banner__detail-label--start-time"
        >
          Start Time
        </td>
        <td
          className="m-block-waiver-banner__detail-value m-block-waiver-banner__detail-value--start-time"
        >
          4:26pm
        </td>
      </tr>
      <tr>
        <td
          className="m-block-waiver-banner__detail-label m-block-waiver-banner__detail-label--end-time"
        >
          End Time
        </td>
        <td
          className="m-block-waiver-banner__detail-value m-block-waiver-banner__detail-value--end-time"
        >
          4:43pm
        </td>
      </tr>
    </tbody>
  </table>
  <div
    className="__react_component_tooltip-304517788 __react_component_tooltip place-bottom type-dark"
    data-id="tooltip"
    id="reason"
  >
    <h3>
      Reason codes
    </h3>
    <ul>
      <li>
        <strong>
          B
        </strong>
         - Manpower
      </li>
      <li>
        <strong>
          C
        </strong>
         - No equipment
      </li>
      <li>
        <strong>
          D
        </strong>
         - Disabled bus
      </li>
      <li>
        <strong>
          E
        </strong>
         - Diverted to other work
      </li>
      <li>
        <strong>
          F
        </strong>
         - Traffic
      </li>
      <li>
        <strong>
          G
        </strong>
         - Accident
      </li>
      <li>
        <strong>
          H
        </strong>
         - Weather
      </li>
      <li>
        <strong>
          I
        </strong>
         - Operator error
      </li>
      <li>
        <strong>
          J
        </strong>
         - Other
      </li>
    </ul>
  </div>
</div>
`;

exports[`BlockWaiverBanner renders a past time waiver 1`] = `
<div
  className="m-block-waiver-banner m-block-waiver-banner--past"
>
  <div
    className="m-block-waiver-banner__header"
  >
    <span
      className="m-block-waiver-banner__alert-icon"
    >
      <svg
        className="c-icon-alert-circle"
        viewBox="-3 -3 54 54"
      >
        <g
          className="c-icon-alert-circle--grey-on-grey"
        >
          <circle
            className="c-icon-alert-circle__outline"
            cx="24"
            cy="24"
            r="27"
          />
          <circle
            className="c-icon-alert-circle__fill"
            cx="24"
            cy="24"
            r="22.59"
          />
          <g
            className="c-icon-alert-circle__exclamation-point"
          >
            <path
              d="m20.39 4.42h7.22a1.81 1.81 0 0 1 1.49 2.11l-1.49 23.22a1.73 1.73 0 0 1 -1.49 1.78h-4.24a1.73 1.73 0 0 1 -1.49-1.78l-1.49-23.22c-.07-1.13.61-2.11 1.49-2.11"
            />
            <path
              d="m21.57 34.54h4.87a1.88 1.88 0 0 1 1.82 1.93v5.17a1.89 1.89 0 0 1 -1.82 1.94h-4.87a1.89 1.89 0 0 1 -1.83-1.94v-5.17a1.88 1.88 0 0 1 1.83-1.93"
            />
          </g>
        </g>
      </svg>
    </span>
    <div
      className="m-block-waiver-banner__title"
    >
      Dispatcher Note - 
      Past
    </div>
  </div>
  <table
    className="m-block-waiver-banner__details"
  >
    <tbody>
      <tr>
        <td
          className="m-block-waiver-banner__detail-label m-block-waiver-banner__reason"
          data-event="click"
          data-for="reason"
          data-tip={true}
        >
          Reason 
          <span
            className="m-block-waiver-banner__reason-icon"
            dangerouslySetInnerHTML={
              Object {
                "__html": "SVG",
              }
            }
          />
        </td>
        <td
          className="m-block-waiver-banner__detail-value"
        >
          E:1106
        </td>
      </tr>
      <tr>
        <td
          className="m-block-waiver-banner__detail-label m-block-waiver-banner__detail-label--start-time"
        >
          Start Time
        </td>
        <td
          className="m-block-waiver-banner__detail-value m-block-waiver-banner__detail-value--start-time"
        >
          3:36pm
        </td>
      </tr>
      <tr>
        <td
          className="m-block-waiver-banner__detail-label m-block-waiver-banner__detail-label--end-time"
        >
          End Time
        </td>
        <td
          className="m-block-waiver-banner__detail-value m-block-waiver-banner__detail-value--end-time"
        >
          3:53pm
        </td>
      </tr>
    </tbody>
  </table>
  <div
    className="__react_component_tooltip-304517788 __react_component_tooltip place-bottom type-dark"
    data-id="tooltip"
    id="reason"
  >
    <h3>
      Reason codes
    </h3>
    <ul>
      <li>
        <strong>
          B
        </strong>
         - Manpower
      </li>
      <li>
        <strong>
          C
        </strong>
         - No equipment
      </li>
      <li>
        <strong>
          D
        </strong>
         - Disabled bus
      </li>
      <li>
        <strong>
          E
        </strong>
         - Diverted to other work
      </li>
      <li>
        <strong>
          F
        </strong>
         - Traffic
      </li>
      <li>
        <strong>
          G
        </strong>
         - Accident
      </li>
      <li>
        <strong>
          H
        </strong>
         - Weather
      </li>
      <li>
        <strong>
          I
        </strong>
         - Operator error
      </li>
      <li>
        <strong>
          J
        </strong>
         - Other
      </li>
    </ul>
  </div>
</div>
`;
